modules/react/common/spec/components.test-d.tsx

```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import {expectTypeOf} from 'expect-type';
+
+import {createComponent, ElementComponent} from '../lib/utils/components';
+
+it('should assign an element-base component as an ElementComponent', () => {
+  const component = createComponent('div')({Component: (props: {foo: 'bar'}) => null});
+  expectTypeOf(component).toEqualTypeOf<ElementComponent<'div', {foo: 'bar'}>>();
+});
+
+it('should add sub-components to the signature', () => {
+  const component = createComponent('div')({
+    Component: (props: {foo: 'bar'}) => null,
+    subComponents: {
+      Foo: 'bar',
+    },
+  });
+  expectTypeOf(component).toEqualTypeOf<ElementComponent<'div', {foo: 'bar'}> & {Foo: number}>();
```

Comment:
This is actually something that `expect-type` fails at. Not sure why. `Foo` is actually a different type here...